### PR TITLE
ip: send an icmp message when TTL is expired

### PR DESCRIPTION
**Pull request by christophefontaine**
Instead of dropping silently the packet, forge and send back an ICMP message "Time To live exceeded".

### CODE REVIEW

File: modules/ip/datapath/ip_ttl_exceeded.c [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+		rte_node_enqueue_x1(graph, node, next, mbuf);
+	}
+
+	return nb_objs;
+}
+
+struct rte_node_register ip_ttl_exceeded_node = {
+	.name = "ip_ttl_exceeded",
+	.process = ip_ttl_exceeded,
```
Comment (rjarry) [inline modules/ip/datapath/ip_ttl_exceeded.c:71]: To be consistent with the other nodes, can you add `_process` to this function name?

File: modules/ip/datapath/ip_ttl_exceeded.c [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	for (uint16_t i = 0; i < nb_objs; i++) {
+		mbuf = objs[i];
+
+		ip = rte_pktmbuf_mtod(mbuf, struct rte_ipv4_hdr *);
+		// No need to check headroom
+		icmp = (struct rte_icmp_hdr *)
+			rte_pktmbuf_prepend(mbuf, sizeof(struct rte_icmp_hdr));
+
+		// Get the local router IP address from the input iface
+		input_iface = ip_output_mbuf_data(mbuf)->input_iface;
```
Comment (rjarry) [inline modules/ip/datapath/ip_ttl_exceeded.c:42]: It looks like you forgot to commit something else :) Nothing sets this input_iface value.

**Review by rjarry — CHANGES_REQUESTED** [review]

File: modules/ip/datapath/ip_ttl_exceeded.c [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+		ip4_addr_t local_ip = ip4_addr_get(input_iface->id)->ip;
+
+		ip_data = ip_local_mbuf_data(mbuf);
+		ip_data->vrf_id = vrf_id;
+		ip_data->src = local_ip;
+		ip_data->dst = ip->src_addr;
+
+		// RFC792 payload size: ip header + 64 bits of original datagram
+		ip_data->len = sizeof(struct rte_icmp_hdr)
+			+ RTE_MIN(sizeof(struct rte_ipv4_hdr) + 8,
```
[the remark this answers is not in the document]
Comment (christophefontaine) [reply modules/ip/datapath/ip_ttl_exceeded.c:53]: according to the rfc, we have to include the original IP header + 64 bits, within the udp/tcp src/dst ports are defined.

File: modules/ip/datapath/ip_ttl_exceeded.c (hunk 1 above)
Comment (christophefontaine) [reply modules/ip/datapath/ip_ttl_exceeded.c:71]: will do.

File: modules/ip/datapath/ip_ttl_exceeded.c (hunk 2 above)
Comment (christophefontaine) [reply modules/ip/datapath/ip_ttl_exceeded.c:42]: indeed !

File: modules/ip/datapath/ip_ttl_exceeded.c [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+#include <gr_ip4_datapath.h>
+#include <gr_log.h>
+#include <gr_mbuf.h>
+
+#include <rte_common.h>
+#include <rte_graph_worker.h>
+#include <rte_icmp.h>
+#include <rte_ip.h>
+
+#define GR_IP_ICMP_TTL_EXCEEDED 11
```
[the remark this answers is not in the document]
Comment (rjarry) [reply modules/ip/datapath/ip_ttl_exceeded.c:16]: Can you put this constant in gr_ip4_datapath?

**Review by rjarry — CHANGES_REQUESTED** [review]

**Review by rjarry — APPROVED** [review]

**Pull request merged**